### ISSUE TITLE
Segmentation fault (core dumped)

### ISSUE THREAD

**Issue by Macleykun** [lowitea/gitlobster#33]
Hello! Thank you for making gitlobster, after getting the import for the groups to work, i was sadly greeted with that it only effectively copies the group (no projects). Your solution looks very promising and being able to pre-migrate everything and only run an update once more (or a few times) when we migrate is fantastic!

I run the binairy as root on rhel 8.7 and calling the help or not providing the required args also shows the expected output. However whenever i supply the needed args i get a Segmentation fault (core dumped). Due to this error i wanted to ask if i could help to debug what the real error is and how we/i can resolve it.

Command used (redacted ofcours)
root@rhel8.7# ./gitlobster --ft="glpat-atoken" --fu="https://sourcegitlab" --bt="glpat-btoken" --bu="https://destgitlab" --dry-run
Segmentation fault (core dumped)

Same happens when i remove the optional arguments. Tried both 1.0 and 1.1, but sadly the same problem occurred.

**Comment by lowitea** [comment]
Hello! Thank you so much for the report!
Currently, binary files are builded only linux systems wiith the x86_64 architecture and OS with musl. I suspect that you don't have musl. If possible, try installing the package via cargo. This is the most reliable option. Or you can try to run through docker. You can also try running Gitlobster with the `-vvvv` option, this will output a detailed log of the command execution.

**Comment by Macleykun** [comment]
> Hello! Thank you so much for the report!
> Currently, binary files are builded only linux with musl and x86_64 architecture. I suspect that you don't have musl. If possible, try installing the package via cargo. This is the most reliable option. Or you can try to run through docker. You can also try running gitlobster with the `-vvvv` option, this will output a detailed log of the command execution.

Aaah I see! I will test this tomorrow
I think bc we have a closed off environment to use the package perhaps
https://pkgs.org/search/?q=Musl
I’ll assume it the musl lib static, but will try to install them all and test, if it doesn’t work I can get the container in and try again.

thank you for your helpful and quick response!

**Comment by Macleykun** [comment]
thanks! the container works great but i somehow had a timeout a few times, but retrying it a third time worked
i only had 443 open, so need to get 80 open aswell as i noticed the cloning is done over http.
This is not a problem for me personally, but i'm curious what you think of detecting that if the remote uses https, to also use cloning through https.

Ofcours ssl should be setup properly on the OS level by the enduser.

**Comment by lowitea** [comment]
I'm glad it worked out!

The default timeout is 30 seconds. I plan to make a flag through which it will be possible to change the timeout.

The mechanism works like this: a list of repositories along with links for cloning is taken from the GitLab API. If you are downloading over http, it means that links with this protocol are obtained from gitlab. You can use the `--download-ssh` flag to download over ssh protocol.

But I can also make a flag in gitlobster with which to force http or https to be selected... Thanks for the idea :)

**Comment by Macleykun** [comment]
> I'm glad it worked out!
> 
> The default timeout is 30 seconds. I plan to make a flag through which it will be possible to change the timeout.
> 
> The mechanism works like this: a list of repositories along with links for cloning is taken from the GitLab API. If you are downloading over http, it means that links with this protocol are obtained from gitlab. You can use the `--download-ssh` flag to download over ssh protocol. In this case, ssh keys of the host user should be added to the gitlab.
> 
> But I can also make a flag in gitlobster with which to force http or https to be selected... Thanks for the idea :)

your welcome! i do want to try the http methode as that seems to work (almost).
Currently, and i hope this is the last obstacle...
![image](https://user-images.githubusercontent.com/26381427/212166597-bd58f645-2cc8-4413-9146-3448a13bbd62.png)

Retrying gives:

Error: git error: error: branch 'remotes/origin/HEAD -> origin/master' not found.

This is what kind of branches are specified:
```
root@dcdd3e81cb83:/tmp/gitlobster/linux/clamav# git branch -avv
* master                  2ece603 [upstream/master] commit message
  remotes/upstream/HEAD   -> upstream/master
  remotes/upstream/master 2ece603 commit message
``` 

We sadly use the oudated primairy name for branches, but this is something i wish to address during the migration hopefully.

You have any idea how i could resolve this? The repo/group(s) aren't made yet on the remote by gitlobster so i'm not sure what to specify/what to do to resolve this.

I do very appricate your help and again your work, i'm certain this will work for sure :) !

**Comment by Macleykun** [comment]
Realized I might have to also create the destination group perhaps, will try this tomorrow and report (and hopefully close haha)

**Comment by lowitea** [comment]
I hope so too :D
But I am very grateful for your feedback, thank you.
You last error looks like a bug :(

You need to try the following:
1. Remove the local out folder (for clean run)
2. Run with flag `-vvvv` and show last 50 lines (check sensitive data)

**Comment by Macleykun** [comment]
I've send the logs to your email specified on your profile.
It did work when i only specified one repo as a test, so it is certain working.

a different question i have is, when i don't specify the --bg= flag. I would assume it goes to the highest/root of the gitlab group, but instead i don't see anything new happening.

Sadly, we have  quite alot of projects underneath a group that's on the highest level, and i was wondering how i could specify to synchronise two top most groups with their subgroups and projects.

**Comment by lowitea** [comment]
> I've send the logs to your email specified on your profile.
> It did work when i only specified one repo as a test, so it is certain working.

Aaah :( I forgot to ask to disable concurrency. Sorry.
Could you please repeat the launch with the additional flag `--concurrency-limit 1`? This does synchronous cloning and arranges the log messages. And you will see a repository with a problem. I'm sure the problem is with a specific repository. Perhaps there is a branch named `HEAD` in the repository, perhaps this is causing this error...

> a different question i have is, when i don't specify the --bg= flag. I would assume it goes to the highest/root of the gitlab group, but instead i don't see anything new happening.

Unfortunately, no. The flag has no default value. If you want to push repositories to another gitlab, you should specify three flags: `--bg`, `--bu`, `--bt`. At the moment, gitlobster cannot push downloaded repositories to the root of another gitlab. But I'm thinking about how to implement it. Now you definitely need to specify the group where the downloaded repositories will be pushed.

**Comment by Macleykun** [comment]
It's ok! It can happen! You can discard the email with attachment then :)
i reran it and it was indeed alot easier to read, infact i think i maybe know what causes it

The repostitory that seems to cause this fail is emptry and never really had a repo to begin with!
By creating a project, choosing blank project, unchecking initilize repo with readme.
Causes this exception where you have a project, but not a repo which indeed has no git info which fails in the end!

What you think :) ? i will probably delete the repo as it hasn't been used in forever, willl update you on this!

**Comment by lowitea** [comment]
I am very glad that I was able to find the problem. Yes, this is really possible. And it definitely needs to be corrected in gitlobster to exclude such repositories. I will set up a ticket for this and will do it in the near future. Thank you so much for helping me find this problem! :)

**Comment by lowitea** [comment]
I was able to reproduce this bug. As you wrote, everything breaks on an empty project, I will definitely fix it :)
By the way, instead of deleting the repository, you can use the `--exclude` flag, if it is more convenient.

**Comment by Macleykun** [comment]
> I am very glad that I was able to find the problem. Yes, this is really possible. And it definitely needs to be corrected in gitlobster to exclude such repositories. I will set up a ticket for this and will do it in the near future. Thank you so much for helping me find this problem! :)

> I was able to reproduce this bug. As you wrote, everything breaks on an empty project, I will definitely fix it :) By the way, instead of deleting the repository, you can use the `--exclude` flag, if it is more convenient.

no worries! It's fun to find bugs and work with devs to resolve them :D 

I'll keep the exclude in mind, i'm more of a clean it up kinda guy and it was from 2017 made by a deleted user :)

The sync worked btw! All repo's are sync'd up now!

One (really) final feature request, is it possible to not have the description say when it was last synce'd by gitlobster? If a user choose to change from the default to not have the description, it's up to them to verify if the sync actually worked (this can be done by checking the return code for example, but that's up to the user ;)).

Thank you again, and i'll keep an eye out on those tickets! Have a great weekend!

**Comment by lowitea** [comment]
> The sync worked btw! All repo's are sync'd up now!

I am very happy about it :)

> One (really) final feature request, is it possible to not have the description say when it was last synce'd by gitlobster? If a user choose to change from the default to not have the description, it's up to them to verify if the sync actually worked (this can be done by checking the return code for example, but that's up to the user ;)).

Sure. I'm surprised I haven't done it yet...

> Thank you again, and i'll keep an eye out on those tickets! Have a great weekend!

And thank you, your feedback was really useful to me. Have a nice weekend :)

**Comment by Macleykun** [comment]
(ok final final, but it's more a suggestion)
I saw in the most verbose logging a few entries were warnings:
```
2023-01-13T17:04:27.207614Z  WARN gitlobster::git: err="Everything up-to-date\n"
2023-01-13T17:04:27.968326Z  WARN gitlobster::git: err="hint: Pulling without specifying how to reconcile divergent branches is\nhint: discouraged. You can squelch this message by running one of the following\nhint: commands sometime before your next pull:\nhint: \nhint:   git config pull.rebase false  # merge (the default strategy)\nhint:   git config pull.rebase true   # rebase\nhint:   git config pull.ff only       # fast-forward only\nhint: \nhint: You can replace \"git config\" with \"git config --global\" to set a default\nhint: preference for all repositories. You can also pass --rebase, --no-rebase,\nhint: or --ff-only on the command line to override the configured default per\nhint: invocation.\nFrom http://source.domain.ext/group/project\n * branch            main       -> FETCH_HEAD\n"
```

I don't think it's an error if everything is up-to-date, in fact i would put it under: INFO if possible.
The second one is good, but perhaps in the container image we can set it?
```
Pulling without specifying how to reconcile divergent branches is
discouraged. You can squelch this message by running one of the following
commands sometime before your next pull:

  git config pull.rebase false  # merge (the default strategy)
  git config pull.rebase true   # rebase
  git config pull.ff only       # fast-forward only

You can replace \"git config\" with \"git config --global\" to set a default
preference for all repositories. You can also pass --rebase, --no-rebase,
or --ff-only on the command line to override the configured default per
invocation.
```
I think the easiest way is to just set:
`git config --global pull.rebase false`
in the docker image, or you prefer a different default config?

**Comment by lowitea** [comment]
> (ok final final, but it's more a suggestion)

Haha :) No problems

> I don't think it's an error if everything is up-to-date, in fact i would put it under: INFO if possible.

This is an error received from git. Git writes this message to stderr so the message gets into warnings. I know about the problem and I'm thinking how to solve it.

> The second one is good, but perhaps in the container image we can set it?

Yea, I agree. That's a good idea, thank you.

> in the docker image, or you prefer a different default config?

I think it's best to install it in every locally downloaded repository. Then it will work without the docker container. Thank you for this remark :)

**Comment by lowitea** [comment]
@Macleykun I have released a new version with correction #37 and the addition of the option `--disable-sync-date` (#41) to disable the output of synchronization time in the description. You can upgrade your docker image to version 1.2.0.

**Comment by Macleykun** [comment]
> @Macleykun I have released a new version with fix #37 and the addition of the option `--disable-sync-date` (#41) to disable the output of synchronization time in the description. You can upgrade your docker image to version 1.2.0.

Thank you for the fixing the bug and adding that feature!
I will however wait with updating, as i technically need to wait for: https://github.com/lowitea/gitlobster/issues/36 (no need to rush/hurry!!!!). While the workaround is ok, i would lose the ability to keep syncing the old environment with the new one (b/c i have to keep moving them which is less ideal).

Again, no need to rush as i also have an annoying nesting issue with the ldap group sync: https://github.com/Adambean/gitlab-ce-ldap-sync . Something i'll have to figure out with myself/a college or maybe the maintainer :).

**Comment by Macleykun** [comment]
I'll close this as in the next (docker) release i'll update to it, and will setup the sync's!
Thank you so much again :) really excited to add the most important part of our project!
